Fixed a bug in EuiCodeBlock docs where import statement example was raised up and evaluated.

1Copenut opened the pull request:
## Summary

EUI end user [reported a bug](https://github.com/elastic/eui/issues/6585#issuecomment-1426335196) in the `EuiCodeBlock` sandbox this weekend. After investigating I determined the sample `import React from 'react'` was being raised up and evaluated by CodeSandbox instead of being rendered as a code block. I ~~changed the block to a general JS object and tested locally to ensure it rendered properly.~~ adjusted the regex logic to ignore the string "React" in code comments.

---
**Before**

<img width="2157" alt="error" src="https://user-images.githubusercontent.com/934879/218495064-97e1fed7-cc49-44b5-a58c-4d7a5e1c57da.png">

---
**After**

<img width="1421" alt="react" src="https://user-images.githubusercontent.com/934879/218535857-7ca1d1b6-b778-4133-a329-ccb8330d45f4.png">

## QA

Remove or strikethrough items that do not apply to your PR.

### General checklist

- [x] Checked in both **light and dark** modes
- [x] Checked in **mobile**
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
- [x] Added or updated [jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests
- [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples
- [x] Checked for **breaking changes** and labeled appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6595/

miukimiu commented:
@1Copenut IMO we shouldn't change the example. But we should fix the regex that transforms the demo code to the `demoJS` tab. 

If we pass a code like the following one:

```js
const jsCode = `/* I'm an example of JS */
import React from 'react';`;
```

When we open the `DemoJS` tab, the regex thinks that `import React from 'react';` that was passed as a string and assigned to `const jsCode` is a real `import`. This is not correct. As we can see in the following screenshot the `const jsCode` is not correct. It doesn't contain the import that was moved to the third line. Consequently, the CodeSandbox fails.

<img width="1133" alt="Screenshot 2023-02-13 at 15 58 37" src="https://user-images.githubusercontent.com/2750668/218508067-798ee966-5487-4516-ba3c-8521c35bc2fb.png">

The regex that generates the `DemoJS` tab can be found here: https://github.com/elastic/eui/blob/main/src-docs/src/components/guide_section/_utils.js. Not sure if it is a lot of work to fix this... 😬  

CC @cee-chen.

1Copenut commented:
> IMO we shouldn't change the example. But we should fix the regex that transforms the demo code to the `demoJS` tab.

Excellent call. I refactored the regex that populates the `remainingImports[]` to ignore matches with "React" string in them. Added a unit test to confirm this behavior.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6595/

cee-chen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -80,7 +79,7 @@ export const renderJsSourceCode = (code) => {
     // ([^]+?)                       - capture any characters (including newlines)
     //  from ('[A-Za-z0-9 -_.@/]*';) - ` from 'someLibrary';` - alphanumeric and certain special characters only
-    /(\/\/.+\n)?import ([^]+?) from ('[A-Za-z0-9 -_.@/]*';)/g,
+    /(\/\/.+\n)?import ((?!React)[^]+?) from ('[A-Za-z0-9 -_.@/]*';)/g,
```
While this works, I'm not a fan because it's not actually targeting the issue we want to solve - which is to say, it's finding an `import` string _within a code snippet_ that isn't actually an import. So if the import snippet in question changed to `import foo from 'bar'`, we'd still have the exact same issue.

I would strongly rather fix the [negative lookbehind](https://javascript.info/regexp-lookahead-lookbehind#lookbehind) to find **any** backtick character before **any** `import` statement and not `React` specifically. You should be able to accomplish this like so:

```suggestion
    /(?<!(`[\s\S]*))(\/\/.+\n)?import ([^]+?) from ('[A-Za-z0-9 -_.@/]*';)/g,
```

Since regexes are incredibly hard to read, we should also document this new addition in the code above line 78:

```js
    // (?<!(`[\s\S]*))               - negative lookbehind ensuring that import statements aren't part of a template literal
    // (\/\/.+\n)?                   - optional preceding comments that must be above specific imports, e.g. // @ts-ignore

cee-chen replied:
There should be no scenario where a real import comes after a template literal backtick - our current eslint rules ensure that all legitimate `import`s are at the top of the file, so this is a fairly safe assertion to make.

cee-chen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -14,7 +14,6 @@ export const renderJsSourceCode = (code) => {
    * Extract React import (to ensure it's always at the top)
    */
   let reactImport = '';
-
```
I'd prefer we not remove newlines at random 😅 can we revert this change please?

cee-chen commented on src-docs/src/components/guide_section/_utils.test.js:
```diff
@@ -314,5 +314,30 @@ describe('renderJsSourceCode', () => {
       );
     });
+
+    it('handles importing React correctly by ignoring code comments', () => {
```
```suggestion
    it('does not handle import statements within template literal backticks', () => {
```

Test itself looks great!

cee-chen commented:
Code changes LGTM - waiting for CI/staging to update and will approve then

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6595/

cee-chen reviewed (approved):
Staging LGTM! Nice attention to detail when catching this bug Trevor!
